core/src/main/kotlin/gropius/util/schema/Schema.kt

```diff
@@ -0,0 +1,183 @@
+package gropius.util.schema
+
+import graphql.schema.validation.InvalidSchemaException
+
+/**
+ * Represents a JSON Type Definition schema.
+ * Modified version of https://github.com/jsontypedef/json-typedef-java/blob/master/src/main/java/com/jsontypedef/jtd/Schema.java
+ */
+class Schema(
+    val nullable: Boolean = false,
+    val metadata: Map<String, Any>? = null,
+    val ref: String? = null,
+    val type: Type? = null,
+    val enum: Set<String>? = null,
+    val elements: Schema? = null,
+    val properties: Map<String, Schema>? = null,
+    val optionalProperties: Map<String, Schema>? = null,
+    val values: Schema? = null,
+    val discriminator: String? = null,
+    val mapping: Map<String, Schema>? = null,
+    val definitions: Map<String, Schema>? = null
+) {
+
+    /**
+     * Ensures the schema is a valid root schema.
+     *
+     * The JSON Type Definition has some rules for the correctness of a schema
+     * that go beyond what can be represented in Java's type system. This method
+     * will verify these rules, such as ensuring that all references have a
+     * corresponding definition, and only valid combinations of JSON Typedef
+     * keywords are used.
+     *
+     * @throws InvalidSchemaException if the schema is not valid
+     */
+    fun verify() {
+        verify(this)
+    }
+
+    /**
+     * Verifies that this schema is valid
+     *
+     * @param root the root schema
+     * @throws InvalidSchemaException if the schema is not valid
+     */
+    private fun verify(root: Schema) {
```

Comment:
unnecessary long, please butcher

Reply:
this is a fork of the original json type definition validator
I would prefer leave this code as close as possible to the original to make reacting to changes in the original simpler, in particular in the validation code